### PR TITLE
Fix sample app: nested paths during decoding, supporting optional values

### LINKED ISSUE #27: Example app does not run step 2 or later
To track the issue here, `swift run` errors out a.t.m.:

```
Swift/ErrorType.swift:200: Fatal error: Error raised at top level: 
Swift.DecodingError.keyNotFound(CodingKeys(stringValue: "price", intValue: nil), 
Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "price", intValue: nil)], 
debugDescription: "No value associated with key CodingKeys(stringValue: \"price\", intValue: nil) (\"price\").", 
underlyingError: nil))
```

Both for 

    swift-driver version: 1.62.3 Apple Swift version 5.7 (swiftlang-5.7.0.123.8 clang-1400.0.29.50)

and

    swift-driver version: 1.45.2 Apple Swift version 5.6.1 (swiftlang-5.6.0.323.66 clang-1316.0.20.12)

### PR DESCRIPTION
**Pull request by DivineDominion**
Fixes #27 

To be honest, I don't quite understand what's going on to make this happen, and what the implications are.

I'm fiddling around with tests to reproduce the problem and catch it there, too.

In the meantime, pointers and feedback would be welcome :)

### CODE REVIEW

**Comment by icanzilb** [comment]
Thanks for catching and fixing this!

One of the PRs that got merged since yesterday must've broken the demo app. And I suspect is my PR here https://github.com/icanzilb/MarkCodable/pull/22 but it's strange that none of the tests broke.

Give me a moment to double-check which code path that the previous PR removed was handling this use case just to make sure we're fixing this for good.

**Comment by rizwankce** [comment]
@icanzilb I just tested it out. It seems CodingPath is empty for `price.price`. We removed Collection type checks & nested containers with this 1845f68ba48606ab6b719a6cec7869772f31dc68 

I could be wrong.

**Comment by DivineDominion** [comment]
Feel free to merge if this works for you 👍

File: Tests/MarkCodableTests/Utility/TestModels.swift [hunk 1]
```diff
@@ -117,3 +117,11 @@ struct Lists: Codable, Equatable {
 struct ListContainer<T: Codable>: Codable {
     let numbers: Array<T>
 }
+
+struct AnimalFarm: Codable, Equatable {
+    let pig: Pig
```
Comment (icanzilb) [inline Tests/MarkCodableTests/Utility/TestModels.swift:122]: @DivineDominion could we please use this opportunity to flex a few more code paths here? Looking at `MarkKeyedDecoding` I can think of 1) adding an optional `Pig` to walk the optional decoding paths 2) adding a optional property inside Pig set to `nil` to walk the decodeNil method.

```
    let pig: Pig
    let optionalPig: Pig?
```

and then

```
struct Pig ... {
  let name: String
  let color: String? 
}
```

File: Sources/MarkCodable/Decoder/MarkKeyedDecoding.swift [hunk 2]
```diff
@@ -80,7 +80,8 @@ struct MarkKeyedDecoding<Key: CodingKey>: KeyedDecodingContainerProtocol {
     }
 
     private func unbox<T: Decodable & StringInitializable>(_ key: Key) throws -> T {
-        guard let value = data[key.stringValue] else {
+        let nestedPath = codingPath + [key]
```
Comment (icanzilb) [inline Sources/MarkCodable/Decoder/MarkKeyedDecoding.swift:83]: This makes very much sense — it was a big oversight we didn't have tests to catch this, thanks for the fix

**Review by icanzilb — CHANGES_REQUESTED** [review]
It looks good and it fixes the bug, the only ask I have is to use the unit test to do few more checks

File: Tests/MarkCodableTests/Utility/TestModels.swift (hunk 1 above)
Comment (DivineDominion) [reply Tests/MarkCodableTests/Utility/TestModels.swift:122]: This is quite tough! (And now the scope of the PR changes accidentally.)

I've devised a test, but I'm not sure which variant is actually expected.

TLDR: @icanzilb do we want to add column headers for columns that only have `nil` values? Or do we want to skip them?

## Testing empty column headers are added ❌ 

This fails:

```swift
    func testSingleOptionalString() throws {
        struct SingleOptionalString: Codable, Equatable {
            var optionalString: String?
        }

        let encoder = MarkEncoder()
        let decoder = MarkDecoder()

        let singleNilMarkdown = """
        |optionalString|
        |--------------|
        |              |
        """
        let nilValue = SingleOptionalString(optionalString: nil)
        XCTAssertEqual(singleNilMarkdown, try encoder.encode(nilValue))
        XCTAssertEqual(nilValue, try decoder.decode(SingleOptionalString.self, string: singleNilMarkdown))

        let singleNonNilMarkdown = """
        |optionalString|
        |--------------|
        |yes           |
        """
        let existingValue = SingleOptionalString(optionalString: "yes")
        XCTAssertEqual(singleNonNilMarkdown, try encoder.encode(existingValue))
        XCTAssertEqual(existingValue, try decoder.decode(SingleOptionalString.self, string: singleNonNilMarkdown))
    }
```

## Same test, without header

Difference is:

```swift
        let singleNilMarkdown = """
        ||
        ||
        ||
        """
```

That does pass, but it's weird: why do we get a table at all, why no separator lines? I'm not sure if this is okay or not.

Comment (icanzilb) [reply Tests/MarkCodableTests/Utility/TestModels.swift:117]: I would add the columns that only have nil values **if we are able to** — this will give chance to humans to enter values in a plain text editor. I didn't want to expand the PR scope with my comment though, I thought it'd be trivial to expand the test 🙈

Do you feel like looking into the `nil` value issue in here? Otherwise we can merge and have a separate issue for this - up to you

Comment (DivineDominion) [reply Tests/MarkCodableTests/Utility/TestModels.swift:117]: @icanzilb When do you think we would not (or should not?) be able to?

Comment (icanzilb) [reply Tests/MarkCodableTests/Utility/TestModels.swift:117]: @DivineDominion I'm not certain the encoder walks all the properties if they are set to `nil`.

Comment (DivineDominion) [reply Tests/MarkCodableTests/Utility/TestModels.swift:117]: @icanzilb I got this working now, fixing #1 in the process (which is similar to what I've posted above)

Turns out we need `*IfPresent` for all primitive types. Yikes. Leaving that for a follow-up.

**Comment by DivineDominion** [comment]
@icanzilb PTAL

**Review by icanzilb — APPROVED** [review]
This looks good — it fixes a few things off the issues list and that's great! 

I'll have to sit down and get a good overview of where we come short while encoding/decoding — e.g. I had no idea the *IfPresent methods were needed at all...

**Comment by DivineDominion** [comment]
@icanzilb Yeah I just noticed via caveman debugging that none of the others were ever called. I do wonder what the default protocol implementations do -- they don't use the generic encoder/decoder, that's for sure 🤔

I believe the next best thing is to duplicate the single optional string test case for all primitive values, and group that in a separate test case file for overview. Or maybe generate the code via Sourcery, if that's you thing :) But to be 100% certain what happens when, we need to cover more permutations, which is quite annoying.

I envy Apple and the fact that they don't test :-----)

**Comment by icanzilb** [comment]
@DivineDominion that's not entirely true, there's plenty of tests and I do intend to "snatch" them https://github.com/MarkCodable/MarkCodable/issues/20
